### DISCUSSION TITLE
Understanding media queries

### DISCUSSION THREAD

**Discussion by BMCwebdev** [facebook/stylex#549] [Q&A] [answered]
Last one for a few days, I'll go spend time with the kiddo and leave you in peace! I hope maybe there some value in having some feedback on how someone is trying to use the library?

Anyhoo, I'm attempting to build a Columns component. I'll keep my example simple here. I'm trying to give users of the library the ability to choose how many columns they want at different `@media` break points. Struggling a bit on this one.

I'm trying to do this without asking consumers of my component library to have to create their own Stylex object to pass back to the component. While I love the option, I would like to keep the interaction between the consumer and the component library via props (variants) as much as possible.

Here is my current code that seems to always overwrite one or the other column style
```tsx
import React from 'react';
import * as stylex from '@stylexjs/stylex';

// real version would have more than three, but also obviously needs to be a per-determined amount
type Columns = '1' | '2' | '3';

interface ColumnsProps {
  columns?: Columns;
  columnsSM?: Columns;
}

const SMALL = '@media (max-width: 420px)';

const columnsBaseStyles = stylex.create({
  base: {
    display: 'grid',
  }
});
const columnsStyles = stylex.create({
  1: {
    gridTemplateColumns: '1fr'
  },
  2: {
    gridTemplateColumns: 'repeat(2, 1fr)'
  },
  3: {
    gridTemplateColumns: 'repeat(3, 1fr)'
  },
});
const columnsStylesSM = stylex.create({
  1: {
    gridTemplateColumns: {
      [SMALL]: '1fr'
    }
  },
  2: {
    gridTemplateColumns: {
      [SMALL]: 'repeat(2, 1fr)'
    }
  },
  3: {
    gridTemplateColumns: {
      [SMALL]: 'repeat(3, 1fr)'
    }
  },
});

const Columns: React.FC<ColumnsProps> = ({
  children,
  columns = '1',
  columnsSM,
}) => {

  return (
    <div {...stylex.props(
      columnsBaseStyles.base,
      columnsStyles[columns],
      columnsSM && columnsStylesSM[columnsSM],
    )}>
      {children}
    </div>
  )
};

export default Columns;
```
Thank you again. Chat soon and have a great weekend!

**Comment by nmn** [comment, marked as answer]
You're supposed to provide *all* values for each CSS property, like `gridTemplateColumns`, in one place. So your two options are:

## Add `default` cases within `columnsStylesSM`.

i.e. something like this:

```tsx
const columnsStylesSM = stylex.create({
  1: {
    gridTemplateColumns: {
      default: '1fr',
      [SMALL]: '1fr'
    }
  },
  2: {
    gridTemplateColumns: {
      default: '1fr',
      [SMALL]: 'repeat(2, 1fr)'
    }
  },
  3: {
    gridTemplateColumns: {
      default: 'repeat(2, 1fr)',
      [SMALL]: 'repeat(3, 1fr)'
    }
  },
});
```

## Use variables to "merge" different breakpoints:

To "merge" values for different breakpoints together, use variables:

```tsx
const columnsStyles = stylex.create({
  1: {
    gridTemplateColumns: '1fr',
    [tokens.columns]: '1fr',
  },
  2: {
    gridTemplateColumns: 'repeat(2, 1fr)',
    [tokens.columns]: 'repeat(2, 1fr)',
  },
  3: {
    gridTemplateColumns: 'repeat(3, 1fr)',
    [tokens.columns]: 'repeat(3, 1fr)',
  },
});
const columnsStylesSM = stylex.create({
  1: {
    gridTemplateColumns: {
      default: tokens.columns,
      [SMALL]: '1fr'
    }
  },
  2: {
    gridTemplateColumns: {
      default: tokens.columns,
      [SMALL]: 'repeat(2, 1fr)'
    }
  },
  3: {
    gridTemplateColumns: {
      default: tokens.columns,
      [SMALL]: 'repeat(3, 1fr)'
    }
  },
});
```

Now, if you apply styles from `columnsStyles`, then `columnsStylesSM` will be able to read the "default" value via the variable.

This is an advanced pattern that should generally not be needed. Mixing and matching different style cases for different media queries is generally an anti-pattern in responsive design. This is a sign that you're trying to create your own "styling system" on top of a styling system. i.e., you're giving too much flexibility and responsibility the users of your design components.

**Reply by BMCwebdev** [reply]
Interesting, I’ve seen this pattern of using a token as a re-definable tool in other discussions, but haven’t tried using it yet. 

But, am I incorrect in thinking that you wouldn’t be able to use this twice with different values on the same page? `tokens.columns` can only hold one value at a time.

I’m curious to hear your thoughts more on the nature of this being an anti-pattern.

I suppose I am handing a sub-styling-system down to the user in a way. However I’m also quite used to seeing things like Layout Components (ie Stack, HStack, Tiles, Columns) in component libraries. The user being able to make choices  (white space, # of columns) at breakpoints or container sizes would be a necessary aspect of the component.

In your work, is layout always baked into larger components? I could see the strength in that, but it would be quite the paradigm shift for the whole team.

**Reply by BMCwebdev** [reply]
_But, am I incorrect in thinking that you wouldn’t be able to use this twice with different values on the same page? tokens.columns can only hold one value at a time._

I _am_ incorrect, just tested it out and I can indeed have more than one component on the page with different values. So is Stylex making new variables for each usage? Not sure I get what is happening here, but for now this is definitely what I was looking for.

**Reply by nmn** [reply]
> In your work, is layout always baked into larger components?

Yes, with minimal options.

> I’m curious to hear your thoughts more on the nature of this being an anti-pattern.
> I’m also quite used to seeing things like Layout Components (ie Stack, HStack, Tiles, Columns) in component libraries.

The specific thing I think is an anti-pattern is making users pick and choose different values for for different media queries. StyleX's decision to put media queries *inside* the value comes from our opinion that they should defined as sets.

Users should choose which responsive layout they want. They shouldn't be mixing and matching different layouts for different screen sizes. That, I believe, is the anti-pattern that leads to fragile layouts that are not properly responsive.

Additionally, a *large* number of layouts can be achieved with modern flexbox and grid layout features without resorting to media queries at all. The pattern above, incentivises the usage of media queries for everything which is inherently more fragile that fluid layouts.

> So is Stylex making new variables for each usage?

No. CSS variables work like context. When you assign values, they are only defined for that element and it's children. Defining it again on a child element will override the value for that element and so on.

Different siblings can have completely disjoint values.

**Reply by BMCwebdev** [reply]
Thank you, @nmn love the thinking. I'm a huge fan of working with CSS have spent a lot of time doing so, but this is the first project I've been able to use CSS variables in (legacy products). Had no idea they worked like context.

Working with Stylex has been an interesting exercise in bubbling issues to the surface. The current Layout Components I am working on is a good example. In general Layout Components are such a hot debate item.

As we build out this DS and component library it will be interesting to see if these really are something we still want or need.

Thanks again for the library, and thanks for the great feedback!
